### PR TITLE
High level api

### PR DESCRIPTION
**Pull request by vlad17**
This commit adds the high-level KeyedModels API, which enables users to train scikit-learn estimators on grouped data.

Testing done via unit tests (note some edge case functionality is not offered / tests are skipped until associated spark JIRAs are resolved).

Some additional refactoring in test_gapply was done to extract common pandas dataframe functionality.

### CODE REVIEW

File: python/spark_sklearn/keyed_models.py [hunk 1]
```diff
@@ -0,0 +1,407 @@
+"""
+Classes for implementing keyed models.
+
+The use case that this addresses is where a client has a dataset with many keys - the distribution
+ of distinct keys amongst the is such that the total number of rows for every key value can be
+contained completely in memory on a single machine.
+
+This assumption is particularly enabling because clients may wish to apply more intricate single-
```
Comment (thunterdb) [inline python/spark_sklearn/keyed_models.py]: I would use useful instead of enabling, but I am not a native speaker :-)

File: python/spark_sklearn/keyed_models.py [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+>>> from pyspark.sql import SparkSession
+>>> from pyspark.sql.functions import udf
+>>> spark = SparkSession.builder.master("local").getOrCreate()
+>>> df = spark.createDataFrame([(user,
+...                              Vectors.dense([i, i ** 2, i ** 3]),
+...                              user + i + 2 * i ** 2 + 3 * i ** 3)
+...                             for user in range(3) for i in range(5)])
+>>> df = df.toDF("key", "features", "y")
+>>> km = KeyedEstimator(sklearnEstimator=LinearRegression(), yCol="y").fit(df)
+>>> printnp = lambda nparr: "[" + ", ".join("{:.2f}".format(x) for x in nparr) + "]"
```
Comment (thunterdb) [inline python/spark_sklearn/keyed_models.py]: this works, but this style is rather foreign in python, I find. Can you used the `def` syntax?

Comment (thunterdb) [inline python/spark_sklearn/keyed_models.py]: and you can combine inside the `lambda` below.

File: python/spark_sklearn/keyed_models.py [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+>>> import numpy as np
+>>> from sklearn.linear_model import LinearRegression
+>>> from pyspark.ml.linalg import Vectors
+>>> from pyspark.sql import SparkSession
+>>> from pyspark.sql.functions import udf
+>>> spark = SparkSession.builder.master("local").getOrCreate()
+>>> df = spark.createDataFrame([(user,
+...                              Vectors.dense([i, i ** 2, i ** 3]),
+...                              user + i + 2 * i ** 2 + 3 * i ** 3)
+...                             for user in range(3) for i in range(5)])
```
Comment (thunterdb) [inline python/spark_sklearn/keyed_models.py]: You should show a few rows, it helps to quickly glance what is inside a dataframe (especially for an example)

File: python/spark_sklearn/keyed_models.py [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+>>> km.keyedModels.select("key", coefs("estimator").alias("lr")).show(truncate=False)
++---+-----------------------------------------+
+|key|lr                                       |
++---+-----------------------------------------+
+|0  |intercept: 0.00 coefs: [1.00, 2.00, 3.00]|
+|1  |intercept: 1.00 coefs: [1.00, 2.00, 3.00]|
+|2  |intercept: 2.00 coefs: [1.00, 2.00, 3.00]|
++---+-----------------------------------------+
+<BLANKLINE>
+>>> input = spark.createDataFrame([(0, Vectors.dense(3, 1, -1))]).toDF("key", "features")
```
Comment (thunterdb) [inline python/spark_sklearn/keyed_models.py]: Instead of inserting a blank line, I would move that to a subsquent paragraph with an explanation. "Now we have produced o model for each key, and we can apply it to test data:"

File: python/spark_sklearn/keyed_models.py [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+from spark_sklearn.group_apply import gapply
+
+class _SparkSklearnEstimatorUDT(UserDefinedType):
+    """
+    SQL user-defined type (UDT) for scikit-learn estimator
+    """
+
+    @classmethod
+    def sqlType(cls):
+        return StringType() if sys.version_info[0] < 3 else BinaryType()
```
Comment (thunterdb) [inline python/spark_sklearn/keyed_models.py]: There is no binary type for 2.x ?

Comment (thunterdb) [inline python/spark_sklearn/keyed_models.py]: Actually I misread: can you comment as to the need the switch internal spark representation between python versions? There is no way to convert to the same (binary?) representation?

File: python/spark_sklearn/keyed_models.py [hunk 6]
```diff
@@ -0,0 +1,407 @@
...
+    (per-key clustering is currently unsupported). Because these scikit-learn estimators
+    all derive from the same base type (yielding the same API), yet have different expectations
+    for what methods should be called and with what arguments, this class enumerates two different
+    types of behavior:
+
+    1. `"transformer"`
+
+        Examples: `sklearn.decomposition.PCA`, `sklearn.cluster.KMeans`
+
+        In this case, the estimator will aggregate the all input features for a given key into a
```
Comment (thunterdb) [inline python/spark_sklearn/keyed_models.py]: input feature vectors

File: python/spark_sklearn/keyed_models.py [hunk 7]
```diff
@@ -0,0 +1,407 @@
...
+        `NxD` data matrix, where `N` is the number of distinct rows for the given key and `D` is the
```
Comment (thunterdb) [inline python/spark_sklearn/keyed_models.py]: distinct?

File: python/spark_sklearn/keyed_models.py [hunk 8]
```diff
@@ -0,0 +1,407 @@
...
+    The input column should be numeric or a vector (else a `TypeError` will be thrown at
+    `fit()`-time). Don't use "estimator" as a column name.
+
+    .. :note: Clustering is *not* supported. For instance, the scikit-learn `KMeans` estimator
+    transforms its data into the cluster-mean-distance space - transform will not generate the
+    column of `KMeans` cluster labels.
+    .. :note: While key-based grouping occurs during training, during the transformation/prediction
+    phase of computation, no such aggregation occurs: the number of rows inputted as test data
+    will be equal to the number of rows outputted.
+    .. :note: `spark.conf.get("spark.sql.retainGroupColumns")` assumed to be `u"true"` for both the
```
Comment (thunterdb) [inline python/spark_sklearn/keyed_models.py]: mention that this is the default value for spark 1.4+ so that users do not freak out about setting a config.

File: python/spark_sklearn/keyed_models.py [hunk 9]
```diff
@@ -0,0 +1,407 @@
...
+    keyed estimator and the keyed model.
+    .. :note: Estimators trained, persisted, and loaded across different sklearn versions are not
```
Comment (thunterdb) [inline python/spark_sklearn/keyed_models.py]: :+1:

File: python/spark_sklearn/keyed_models.py [hunk 10]
```diff
@@ -0,0 +1,407 @@
...
+    guarnteed to work.
+    """
+
+    paramSpecs = {
```
Comment (thunterdb) [inline python/spark_sklearn/keyed_models.py]: what is this?

File: python/spark_sklearn/keyed_models.py [hunk 11]
```diff
@@ -0,0 +1,407 @@
...
+        """For all instances, the ordered list of `keyCols` determine the set of groups which each
+        `sklearnEstimator` is applied to.
+
+        For every unique `keyCols` value, the remaining columns are aggregated and used to train the
+        scikit-learn estimator.
+
+        If `yCol` is specified, then this is assumed to be of `"predictor"` type, else a
+        `"transformer"`.
+
+        :param sklearnEstimator: An instance of a sklearn estimator, with parameters configured
```
Comment (thunterdb) [inline python/spark_sklearn/keyed_models.py]: scikit-learn

File: python/spark_sklearn/keyed_models.py [hunk 12]
```diff
@@ -0,0 +1,407 @@
...
+        as desired for each user.
+        :param keyCols: Key column names list used to group data to which models are applied, where
+        implies order lexicographical importance.
+        :param xCol: Name of column of input features used for training and
+        transformation/prediction.
```
Comment (thunterdb) [inline python/spark_sklearn/keyed_models.py]: mention that it sohuld be a vector

File: python/spark_sklearn/keyed_models.py [hunk 13]
```diff
@@ -0,0 +1,407 @@
...
+        :param yCol: Optional. Specifies name of label column for regression or classification
```
Comment (thunterdb) [inline python/spark_sklearn/keyed_models.py]: Instead of optional, I would say something like (Required for predictors)`

File: python/spark_sklearn/keyed_models.py [hunk 14]
```diff
@@ -0,0 +1,407 @@
...
+
+        :raise ValueError: if `sklearnEstimator` is `None`.
+        :raise ValueError: if `sklearnEstimator` does not derive from `sklearn.base.BaseEstimator`.
+        :raise ValueError: if `keyCols` is empty.
+        :raise ValueError: if any column has the name `"estimator"` (
+        :raise ValueError: if reflection checks indicate that parameter estimator is not equipped
+        with a `fit()` method or an appropriate transformation/prediction method.
+        """
+        if sklearnEstimator is None:
+            raise ValueError("sklearnEstimator should not be unspecified or None")
```
Comment (thunterdb) [inline python/spark_sklearn/keyed_models.py]: can you not unuse an affirmative style?

File: python/spark_sklearn/keyed_models.py [hunk 15]
```diff
@@ -0,0 +1,407 @@
...
+        if not isinstance(sklearnEstimator, sklearn.base.BaseEstimator):
+            raise ValueError("sklearnEstimator should be an sklearn.base.BaseEstimator")
+        if len(keyCols) == 0:
+            raise ValueError("keyCols should not be empty")
+        if "estimator" in keyCols + [xCol, yCol]:
+            raise ValueError("keyCols should not contain a column named \"estimator\"")
+
+        for paramName, paramSpec in KeyedEstimator.paramSpecs.items():
```
Comment (thunterdb) [inline python/spark_sklearn/keyed_models.py]: put a comment and an underscore or something on `paramSpec` if this your own making.

File: python/spark_sklearn/keyed_models.py [hunk 16]
```diff
@@ -0,0 +1,407 @@
...
+            setattr(self, paramName, Param(Params._dummy(), paramName, paramSpec["doc"]))
+        super(KeyedEstimator, self).__init__()
```
Comment (thunterdb) [inline python/spark_sklearn/keyed_models.py]: not before? It is usually the first line

File: python/spark_sklearn/keyed_models.py [hunk 17]
```diff
@@ -0,0 +1,407 @@
...
+        keyCols = self.getOrDefault("keyCols")
+        xCol = self.getOrDefault("xCol")
+        yCol = self.getOrDefault("yCol")
+        isLabelled = yCol is not None
+        assert isLabelled == (self.getOrDefault("estimatorType") == "predictor")
+        _validateXCol(dataset.schema, xCol)
+
+        cols = keyCols[:]
+        cols.append(xCol)
+        if isLabelled: cols.append(yCol)
```
Comment (thunterdb) [inline python/spark_sklearn/keyed_models.py]: I checked the pyspark code base and we do not have such one-liners. New line please.

File: python/spark_sklearn/keyed_models.py [hunk 18]
```diff
@@ -0,0 +1,407 @@
...
+        grouped = projected.groupBy(*keyCols)
+        estimator = self.getOrDefault("sklearnEstimator")
+
+        # Potential optimization: broadcast estimator
+
+        import pandas as pd
+        def fitEstimator(_, pandasDF):
+            X = _prepareXCol(pandasDF[xCol], oneDimensional)
+            y = pandasDF[yCol].values if isLabelled else None
+            pandasDF = None
```
Comment (thunterdb) [inline python/spark_sklearn/keyed_models.py]: unused

File: python/spark_sklearn/keyed_models.py [hunk 19]
```diff
@@ -0,0 +1,407 @@
...
+contained completely in memory on a single machine.
+
+This assumption is particularly enabling because clients may wish to apply more intricate single-
+machine models (such as a scikit-learn estimator) to every user.
+
+The API provided here generalizes the scikit-learn estimator interface to the Spark ML one; in
+particular, it allows clients to train their scikit-learn estimators in parallel over a grouped
+and aggregated dataframe.
+
+>>> import numpy as np
```
Comment (vlad17) [inline python/spark_sklearn/keyed_models.py]: unused; will rm.

File: python/spark_sklearn/keyed_models.py (hunk 1 above)
Comment (vlad17) [inline python/spark_sklearn/keyed_models.py]: I think the use is appropriate here.

File: python/spark_sklearn/keyed_models.py (hunk 3 above)
Comment (vlad17) [inline python/spark_sklearn/keyed_models.py]: done

File: python/spark_sklearn/keyed_models.py (hunk 2 above)
Comment (vlad17) [inline python/spark_sklearn/keyed_models.py]: done

File: python/spark_sklearn/keyed_models.py (hunk 4 above)
Comment (vlad17) [inline python/spark_sklearn/keyed_models.py]: The blank line has to be there to match `show()` output, but I will add a paragraph interjecting with your explanation. Done

File: python/spark_sklearn/keyed_models.py (hunk 18 above)
Comment (thunterdb) [inline python/spark_sklearn/keyed_models.py]: if you want to drop a variable, use `del`

File: python/spark_sklearn/keyed_models.py [hunk 20]
```diff
@@ -0,0 +1,407 @@
...
+        import pandas as pd
+        def fitEstimator(_, pandasDF):
+            X = _prepareXCol(pandasDF[xCol], oneDimensional)
+            y = pandasDF[yCol].values if isLabelled else None
+            pandasDF = None
+
+            estimatorClone = sklearn.base.clone(estimator)
+            estimatorClone.fit(X, y)
+            pickled = pickle.dumps(estimatorClone)
+            estimatorClone = None
```
Comment (thunterdb) [inline python/spark_sklearn/keyed_models.py]: same thing here

File: python/spark_sklearn/keyed_models.py [hunk 21]
```diff
@@ -0,0 +1,407 @@
...
+
+    @keyword_only
+    def __init__(self, keyCols=None, xCol=None, outputCol=None, yCol=None, estimatorType=None,
+                 keyedSklearnEstimators=None, outputType=None):
+        """Used by :class:`KeyedEstimator` to generate a :class:`KeyedModel`. Not intended for
+        external use."""
+
+        assert (estimatorType == "predictor") == (yCol is not None)
+        assert estimatorType == "transformer" or estimatorType == "predictor"
+        def implies(a, b): return not a or b
```
Comment (thunterdb) [inline python/spark_sklearn/keyed_models.py]: 2 lines

File: python/spark_sklearn/keyed_models.py [hunk 22]
```diff
@@ -0,0 +1,407 @@
...
+        assert implies(estimatorType == "transformer", outputType == Vector.__UDT__)
+        assert len(keyCols) > 0
+        assert set(keyedSklearnEstimators.columns) == (set(keyCols) | set(["estimator"]))
+
+        for paramName, paramSpec in KeyedModel.paramSpecs.items():
+            setattr(self, paramName, Param(Params._dummy(), paramName, paramSpec["doc"]))
+        super(KeyedModel, self).__init__()
```
Comment (thunterdb) [inline python/spark_sklearn/keyed_models.py]: same comment

File: python/spark_sklearn/keyed_models.py (hunk 16 above)
Comment (thunterdb) [inline python/spark_sklearn/keyed_models.py]: Ok I understand why. Please put a comment explaining why you need to depart from python conventions, it is non-obvious that the base class performs reflective introspection unless you know the pyspark code.

Comment (thunterdb) [inline python/spark_sklearn/keyed_models.py]: Actually, we should probably refactor that part in pyspark to be lazy.

File: python/spark_sklearn/keyed_models.py (hunk 10 above)
Comment (vlad17) [inline python/spark_sklearn/keyed_models.py]: Admittedly weird, but it seemed like the most convenient and appropriate for containing parameter values: I need a collection of the parameter names, a mapping from param names to their documentation, and a mapping from them to their defaults. I can make everything a separate attribute, but it'd make for a lot of attributes. I think that'd be cumbersome to read through, but your call.

File: python/spark_sklearn/keyed_models.py [hunk 23]
```diff
@@ -0,0 +1,407 @@
...
+        # we don't have a guarantee that they'll fit in memory.
+
+        # Potential optimization: broadcast estimator
+
+        shouldPredict = self.getOrDefault("estimatorType") == "predictor"
+        oneDimensional = _isOneDimensional(dataset.schema, xCol)
+        if shouldPredict:
+            cast = KeyedModel._sql_types[type(outputType)]
+        else:
+            assert outputType == Vector.__UDT__
```
Comment (thunterdb) [inline python/spark_sklearn/keyed_models.py]: `, outputType`

File: python/spark_sklearn/keyed_models.py (hunk 11 above)
Comment (vlad17) [inline python/spark_sklearn/keyed_models.py]: done

File: python/spark_sklearn/keyed_models.py [hunk 24]
```diff
@@ -0,0 +1,407 @@
...
+            # The closure of applyEstimator() doesn't know ahead of time that we won't use
+            # the cast value, so it tries to serialize it.
+            cast = None
```
Comment (thunterdb) [inline python/spark_sklearn/keyed_models.py]: I think `del` should work as well?

File: python/spark_sklearn/keyed_models.py (hunk 12 above)
Comment (vlad17) [inline python/spark_sklearn/keyed_models.py]: not necessary.

File: python/spark_sklearn/test_utils.py [hunk 25]
```diff
@@ -66,3 +67,22 @@ def list2csr(x):
...
+        converted = pdDF.apply(lambda col: col.apply(convert if convert else lambda x: x))
+        ordered = converted.sort_values(sortby if sortby else pdDF.columns.tolist())
+        # We need to drop the index after sorting because pandas remembers the pre-sort
+        # permutation in the old index. This would trigger a failure if we were to compare
+        # differently-ordered dataframes, even if they had the same sorted content.
+        unindexed = ordered.reset_index(drop=True)
+        return unindexed
+    actual = normalize(actual)
+    expected = normalize(expected)
+    pd.util.testing.assert_almost_equal(actual, expected)
```
Comment (thunterdb) [inline python/spark_sklearn/test_utils.py]: I am surprised there is no way to adjust the precision here

File: python/spark_sklearn/tests/test_keyed_models.py [hunk 26]
```diff
@@ -0,0 +1,237 @@
...
+from spark_sklearn.test_utils import fixtureReuseSparkSession, assertPandasAlmostEqual
+
+def _sortByComponentWeight(pca):
+    zipped = zip(pca.components_, pca.explained_variance_ratio_)
+    ordered = sorted(zipped, key=lambda x: x[1])
+    return tuple(np.array(unzipped) for unzipped in zip(*ordered))
+
+def _assertPandasAlmostEqual(actual, expected, sortby):
+    def convert_estimators(x): # note convertion makes estimators invariant to training order.
+        if isinstance(x, SparkSklearnEstimator): x = x.estimator
```
Comment (thunterdb) [inline python/spark_sklearn/tests/test_keyed_models.py]: 2 lines

File: python/spark_sklearn/tests/test_keyed_models.py [hunk 27]
```diff
@@ -0,0 +1,237 @@
...
+            return _sortByComponentWeight(x)
+        return x
+    assertPandasAlmostEqual(actual, expected, convert=convert_estimators, sortby=sortby)
+
+@fixtureReuseSparkSession
+class KeyedModelTests(unittest.TestCase):
+
+    class _CustomTransformer(sklearn.base.BaseEstimator):
+        def fit(X, y=None):
+            pass
```
Comment (thunterdb) [inline python/spark_sklearn/tests/test_keyed_models.py]: very complex transform...

File: python/spark_sklearn/keyed_models.py (hunk 5 above)
Comment (vlad17) [inline python/spark_sklearn/keyed_models.py]: Comment added. Conversion seemed a bit hacky to me. It's not like we'd ever support mixing python2 and python3 procs b/w driver and workers.

File: python/spark_sklearn/keyed_models.py (hunk 6 above)
Comment (vlad17) [inline python/spark_sklearn/keyed_models.py]: RESPONSE REQUESTED: vectors not necessary if 1D. I'll leave comment as-is unless you tell me otherwise.

File: python/spark_sklearn/keyed_models.py (hunk 7 above)
Comment (vlad17) [inline python/spark_sklearn/keyed_models.py]: reworded

File: python/spark_sklearn/keyed_models.py (hunk 8 above)
Comment (vlad17) [inline python/spark_sklearn/keyed_models.py]: done

File: python/spark_sklearn/keyed_models.py [hunk 28]
```diff
@@ -0,0 +1,407 @@
...
+    .. :note: Clustering is *not* supported. For instance, the scikit-learn `KMeans` estimator
+    transforms its data into the cluster-mean-distance space - transform will not generate the
+    column of `KMeans` cluster labels.
+    .. :note: While key-based grouping occurs during training, during the transformation/prediction
+    phase of computation, no such aggregation occurs: the number of rows inputted as test data
+    will be equal to the number of rows outputted.
+    .. :note: `spark.conf.get("spark.sql.retainGroupColumns")` assumed to be `u"true"` for both the
+    keyed estimator and the keyed model.
+    .. :note: Estimators trained, persisted, and loaded across different sklearn versions are not
+    guarnteed to work.
```
Comment (vlad17) [inline python/spark_sklearn/keyed_models.py]: guarantee

File: python/spark_sklearn/keyed_models.py [hunk 29]
```diff
@@ -0,0 +1,407 @@
...
+        For every unique `keyCols` value, the remaining columns are aggregated and used to train the
+        scikit-learn estimator.
+
+        If `yCol` is specified, then this is assumed to be of `"predictor"` type, else a
+        `"transformer"`.
+
+        :param sklearnEstimator: An instance of a sklearn estimator, with parameters configured
+        as desired for each user.
+        :param keyCols: Key column names list used to group data to which models are applied, where
+        implies order lexicographical importance.
```
Comment (vlad17) [inline python/spark_sklearn/keyed_models.py]: done

File: python/spark_sklearn/keyed_models.py (hunk 13 above)
Comment (vlad17) [inline python/spark_sklearn/keyed_models.py]: done

File: python/spark_sklearn/keyed_models.py (hunk 14 above)
Comment (vlad17) [inline python/spark_sklearn/keyed_models.py]: RESPONSE REQUESTED: I don't understand what you mean here

File: python/spark_sklearn/keyed_models.py (hunk 15 above)
Comment (vlad17) [inline python/spark_sklearn/keyed_models.py]: RESPONSE REQUESTED: what do you mean by this?

File: python/spark_sklearn/keyed_models.py (hunk 16 above)
Comment (vlad17) [inline python/spark_sklearn/keyed_models.py]: superclass needs the Param attributes pre-set. Comment added.

File: python/spark_sklearn/keyed_models.py (hunk 17 above)
Comment (vlad17) [inline python/spark_sklearn/keyed_models.py]: done

File: python/spark_sklearn/keyed_models.py (hunk 18 above)
Comment (vlad17) [inline python/spark_sklearn/keyed_models.py]: RESPONSE REQUESTED: what do you mean? I'm saying I don't need pandasDF anymore. `del`? Avoid the optimization altogether (also below too, then?)

File: python/spark_sklearn/tests/test_keyed_models.py [hunk 30]
```diff
@@ -0,0 +1,237 @@
...
+
+        def flattenAndConvertNumpy(x):
+            return [Vectors.dense(i) if isinstance(i, np.ndarray) else i
+                    for i in chain.from_iterable(x)]
+
+        inputDF = pd.DataFrame.from_dict(
+            {k: [i for i in range(NUSERS) for _ in range(nExamplesPerUser(i))] for k in keyCols})
+        inputDF[xCol] = flattenAndConvertNumpy(features)
+        inputDF["useless"] = flattenAndConvertNumpy(useless)
+        if labels: inputDF[kwargs["yCol"]] = flattenAndConvertNumpy(labels)
```
Comment (thunterdb) [inline python/spark_sklearn/tests/test_keyed_models.py]: 2 lines

File: python/spark_sklearn/keyed_models.py (hunk 24 above)
Comment (vlad17) [inline python/spark_sklearn/keyed_models.py]: No, that'd remove it from the context.  I'm not trying to delete anything, but rather *give* it an explicit (though trivial value) so it *can* be serialized.

File: python/spark_sklearn/test_utils.py (hunk 25 above)
Comment (vlad17) [inline python/spark_sklearn/test_utils.py]: There is - 5 digits is just the default.

File: python/spark_sklearn/tests/test_keyed_models.py [hunk 31]
```diff
@@ -0,0 +1,237 @@
...
+        inputDF = self.spark.createDataFrame(inputDF)
+
+        ke = KeyedEstimator(**kwargs)
+        km = ke.fit(inputDF)
+
+        actualDF = km.keyedModels.toPandas()
+        _assertPandasAlmostEqual(actualDF, expectedDF, keyCols)
+
+        # Test users with different amounts of points.
+        nTestPerUser = lambda i: NUSERS // 4 if i < NUSERS // 2 else NUSERS * 3 // 4
```
Comment (thunterdb) [inline python/spark_sklearn/tests/test_keyed_models.py]: do you have to `import division` from `__future__` for that to work?

File: python/spark_sklearn/tests/test_keyed_models.py [hunk 32]
```diff
@@ -0,0 +1,237 @@
...
+        expectedOutput = map(makeOutput, localEstimators, Xs)
+        expectedDF = inputDF.copy(deep=True)
+        expectedDF[outputCol] = flattenAndConvertNumpy(expectedOutput)
+
+        inputDF = self.spark.createDataFrame(inputDF)
+        actualDF = km.transform(inputDF).toPandas()
+
+        _assertPandasAlmostEqual(actualDF, expectedDF, keyCols + ["useless"])
+
+    NDIM = 5
```
Comment (thunterdb) [inline python/spark_sklearn/tests/test_keyed_models.py]: bring all these definitions to the top of the class, they are easier to spot

File: python/spark_sklearn/tests/test_keyed_models.py (hunk 27 above)
Comment (vlad17) [inline python/spark_sklearn/tests/test_keyed_models.py]: The paper's in the works.

File: python/spark_sklearn/tests/test_keyed_models.py (hunk 26 above)
Comment (vlad17) [inline python/spark_sklearn/tests/test_keyed_models.py]: done

File: python/spark_sklearn/keyed_models.py (hunk 23 above)
Comment (vlad17) [inline python/spark_sklearn/keyed_models.py]: RESPONSE REQUESTED: huh?

File: python/spark_sklearn/keyed_models.py (hunk 22 above)
Comment (vlad17) [inline python/spark_sklearn/keyed_models.py]: same response

File: python/spark_sklearn/keyed_models.py (hunk 21 above)
Comment (vlad17) [inline python/spark_sklearn/keyed_models.py]: done

File: python/spark_sklearn/tests/test_keyed_models.py (hunk 32 above)
Comment (vlad17) [inline python/spark_sklearn/tests/test_keyed_models.py]: RESPONSE REQUESTED: "all these definitions" = are you expecting something other than `NDIM` moved up?

File: python/spark_sklearn/tests/test_keyed_models.py (hunk 31 above)
Comment (vlad17) [inline python/spark_sklearn/tests/test_keyed_models.py]: No, that makes `/` behave the new way, `//` has always been the same.

File: python/spark_sklearn/tests/test_keyed_models.py (hunk 30 above)
Comment (vlad17) [inline python/spark_sklearn/tests/test_keyed_models.py]: done

File: python/spark_sklearn/tests/test_keyed_models.py (hunk 32 above)
Comment (vlad17) [inline python/spark_sklearn/tests/test_keyed_models.py]: done

File: python/spark_sklearn/keyed_models.py (hunk 6 above)
Comment (vlad17) [inline python/spark_sklearn/keyed_models.py]: done

File: python/spark_sklearn/keyed_models.py (hunk 10 above)
Comment (vlad17) [inline python/spark_sklearn/keyed_models.py]: (made into private)

File: python/spark_sklearn/keyed_models.py (hunk 14 above)
Comment (vlad17) [inline python/spark_sklearn/keyed_models.py]: done

File: python/spark_sklearn/keyed_models.py (hunk 15 above)
Comment (vlad17) [inline python/spark_sklearn/keyed_models.py]: done

File: python/spark_sklearn/keyed_models.py (hunk 18 above)
Comment (vlad17) [inline python/spark_sklearn/keyed_models.py]: done

File: python/spark_sklearn/keyed_models.py (hunk 20 above)
Comment (vlad17) [inline python/spark_sklearn/keyed_models.py]: done

File: python/spark_sklearn/keyed_models.py (hunk 23 above)
Comment (vlad17) [inline python/spark_sklearn/keyed_models.py]: done

**Comment by thunterdb** [comment]
It looks good, thanks. Merging.
